### PR TITLE
:bug: Fix deletion for Applications assessment/review 

### PR DESCRIPTION
**Pull request by gildub**
Fix regression to not display "Discard assessment" kebab option.
Fix error when discarding application's assessment/review when there was only a review.

Resolves https://issues.redhat.com/browse/MTA-282

### CODE REVIEW

**Comment by ibolton336** [comment]
image
Still seeing a 404 when discarding the review/assessment. When an assessment is not present, we are still trying to remove it I think.

**Review by ibolton336 — CHANGES_REQUESTED** [review]
See comments

**Comment by ibolton336** [comment]
Looks like we need a deeper check on line https://github.com/konveyor/tackle2-ui/pull/715/files#diff-71caaa778c3011862397845daf5242f4c371d905054ff89f4df98add9e2b6f91R522 as we have done elsewhere for the new empty assessment query default value. Looking good otherwise, though.

File: client/src/app/queries/assessments.ts [hunk 1]
```diff
@@ -48,3 +34,10 @@ export const useFetchApplicationAssessments = (
       queryResultsByAppId[id].error as AxiosError | undefined,
   };
 };
+
+export const useDeleteAssessmentMutation = () => {
```
Comment (ibolton336) [inline client/src/app/queries/assessments.ts:38]: Why the change in syntax here? 
How about 

```

export const useDeleteAssessmentMutation = () => {
  return useMutation(deleteAssessment, {
    onError: (onError) => onError,
  });
};
```

**Review by ibolton336 — CHANGES_REQUESTED** [review]
image


The app table isn't reflecting the deletion of an assessment/review. I have to refresh the page to see the changes.

**Comment by gildub** [comment]
@ibolton336, one of the invalidateQueries had the wrong key, but it takes few seconds to get the cache to be refreshed.
So I added a refetch to cover it.

**Comment by ibolton336** [comment]
@gildub It still isn't refetching applications. Need a hard refresh to see the updated state

**Comment by gildub** [comment]
@ibolton336, latest makes it working smoothly.

**Comment by ibolton336** [comment]
@gildub Latest still isn't refreshing the application state for me.

**Comment by ibolton336** [comment]
It is not consistently working for me @gildub

File: client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx [hunk 2]
```diff
@@ -499,39 +506,61 @@ export const ApplicationsTable: React.FC = () => {
...
+          title: t("toastr.success.reviewDiscarded", {
             application: application.name,
           }),
           variant: "success",
         });
-
-        queryClient.invalidateQueries([assessmentsQueryKey]);
         queryClient.invalidateQueries([reviewsQueryKey]);
+        queryClient.invalidateQueries([ApplicationsQueryKey]);
+      }
```
Comment (ibolton336) [inline client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx:542]: @gildub We need another fetchApplications() call here.

**Review by ibolton336 — CHANGES_REQUESTED** [review]

**Review by ibolton336 — CHANGES_REQUESTED** [review]

File: client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx [hunk 3]
```diff
@@ -499,39 +506,63 @@ export const ApplicationsTable: React.FC = () => {
...
+  const {
+    mutate: deleteAssessment,
+    isError: isDeleteAssessmentError,
+    error: deleteAssessmentError,
+  } = useDeleteAssessmentMutation();
+
+  const discardAssessmentAndReview = (application: Application) => {
+    if (application.review?.id) {
+      deleteReview(application.review.id!);
+      if (isDeleteReviewError) {
```
Comment (ibolton336) [inline client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx:528]: This code needs to be handled in an async way. Preferably a callback like we have done elsewhere with react-query.

File: client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx [hunk 4]
```diff
@@ -499,39 +506,63 @@ export const ApplicationsTable: React.FC = () => {
...
         fetchApplications();
-      })
-      .catch((error) => {
+      }
+    }
+
+    const assessment = getApplicationAssessment(application.id!);
+    if (assessment && assessment.id) {
+      deleteAssessment(assessment.id);
+      if (isDeleteAssessmentError) {
```
Comment (ibolton336) [inline client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx:549]: Needs a callback for the same reasons above.

**Review by ibolton336 — CHANGES_REQUESTED** [review]

File: client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx [hunk 5]
```diff
@@ -499,39 +508,55 @@ export const ApplicationsTable: React.FC = () => {
...
+    onDeleteReviewSuccess,
+    onDeleteError
+  );
+
+  const { mutate: deleteAssessment } = useDeleteAssessmentMutation(
+    onDeleteAssessmentSuccess,
+    onDeleteError
+  );
+
+  const discardAssessmentAndReview = (application: Application) => {
```
Comment (ibolton336) [inline client/src/app/pages/applications/applications-table-assessment/applications-table-assessment.tsx:552]: Looking good with the fixes. I think QE will probably file another issue re: the notification for delete review & assessment will only show whichever of these api calls finishes first. This may require further rework to add a new mutation that wraps both the review & assessment delete.

**Review by ibolton336 — APPROVED** [review]
LGTM  - nicely done.

File: client/src/app/queries/reviews.ts [hunk 6]
```diff
@@ -27,3 +28,24 @@ export const useFetchReviews = (): IReviewFetchState => {
...
+}
+
+export const useDeleteReviewMutation = (
+  onSuccess: (name: string) => void,
+  onError: (err: AxiosError) => void
+) => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (args: IReviewMutation) => deleteReview(args.id),
```
Comment (ibolton336) [inline client/src/app/queries/reviews.ts:44]: No need to create an async wrapper here.

File: client/src/app/queries/assessments.ts [hunk 7]
```diff
@@ -48,3 +39,23 @@ export const useFetchApplicationAssessments = (
...
+}
+
+export const useDeleteAssessmentMutation = (
+  onSuccess: (name: string) => void,
+  onError: (err: AxiosError) => void
+) => {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (args: IAssessementMutation) => deleteAssessment(args.id),
```
Comment (ibolton336) [inline client/src/app/queries/assessments.ts:54]: No need to create an async wrapper here.
